src/Adapter/Product/CommandHandler/GenerateProductCombinationsHandler.php

```diff
@@ -0,0 +1,194 @@
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\CommandHandler;
+
+use Combination;
+use Db;
+use Doctrine\DBAL\Connection;
+use Exception;
+use Iterator;
+use PrestaShop\PrestaShop\Adapter\Product\AbstractProductHandler;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Command\GenerateProductCombinationsCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\CommandHandler\GenerateProductCombinationsHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CannotAddCombinationException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CombinationException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\ValueObject\CombinationId;
+use PrestaShop\PrestaShop\Core\Product\Generator\CombinationGeneratorInterface;
+use PrestaShopException;
+use Product;
+use SpecificPriceRule;
+
+/**
+ * Handles @see GenerateProductCombinationsCommand using legacy object model
+ */
+final class GenerateProductCombinationsHandler extends AbstractProductHandler implements GenerateProductCombinationsHandlerInterface
+{
+    /**
+     * @var CombinationGeneratorInterface
+     */
+    private $combinationGenerator;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var Db
+     */
+    private $dbInstance;
+
+    /**
+     * @param CombinationGeneratorInterface $combinationGenerator
+     * @param Connection $connection
+     */
+    public function __construct(
+        CombinationGeneratorInterface $combinationGenerator,
+        Connection $connection
+    ) {
+        $this->combinationGenerator = $combinationGenerator;
+        $this->connection = $connection;
+        $this->dbInstance = Db::getInstance();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @todo: multistore
+     */
+    public function handle(GenerateProductCombinationsCommand $command): array
+    {
+        $product = $this->getProduct($command->getProductId());
+        $generatedCombinations = $this->combinationGenerator->generate($command->getGroupedAttributeIds());
+
+        SpecificPriceRule::disableAnyApplication();
```

Comment:
Could you explain with a comment why this is necessary please? I get that it's to improve performance but some details might help to understand what and when it disables exactly
If I get it right the rules are applied at each product/combination addition?

Reply:
Im not really sure, why, but it was like that in legacy. Ill dig deeper and try to explain it.

Reply:
As you explained yourself, Combination->add() calls SpecificPriceRulle->applyAllRules(). When $rules_application_enable = false it avoids to performing specificPriceRule application on every new combination.